### DISCUSSION TITLE
Welcome to MiniDexed Discussions!

### DISCUSSION THREAD

**Discussion by probonopd** [probonopd/MiniDexed#14] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by positionhigh** [comment]
Hi there and gratulations to your MiniDexed Project - this really looks promising.
I am working on the gui/sequencer/tracker part for a new microdexed generation on the teensy platform, including a small tft and a touchscreen (and also, as before, 2 rotary encoders). I am not experienced on the raspberry/circle platform but i do have several pi devices available. You can see some (outdated) screenshots on my project site.
https://codeberg.org/positionhigh/MicroDexed-touch
Let me know if any part of that is interesting to your project - i think many things could be ported since you already have a text based HD44780 display in your code and we share the same "synth engine".

**Reply by probonopd** [reply]
Hello @positionhigh and thanks for reaching out.

Your project has many things I am dreaming of (more than one Dexed instance, effects, sequencer, mixer,...) :+1: but unfortunately I am lacking the hardware to try it out.

MiniDexed is running on Raspberry Pi Zero, 1, 2, 3, and 4 and we try to keep all other hardware as optional as possible, in order to allow as many people as possible to join the fun, and allow to get started with a "naked" Pi. 

For at least the Pi 2 and newer I think the hardware should be capable of running multiple instances and effects. The next objective would be to run mulitple Dexed instances and TX802 style performances and effects like delay, chrous, and reverb.

So if you see some features that you could help port over, that'd be tremendous. As you probably have noticed, the developers of Circle and MicroDexed are both active in this project and have been extremely helpful.

**Comment by positionhigh** [comment]
Hello probonopd,

i have about no knowledge about Circle but i am over a year involved in microdexed so i believe i might know some things about parts of the code and its structure. Everything is deepy interwoven with the teensy audio library, so getting the core to run in circles is really astonishing to me - we also thought about making that move to Circles on the Pi but my development skills are not enough for that task and the main developer would have liked to do it but does not have of the free time to make this possible.
 
i fully understand your path to make as much as possible available with the bare Pi. But sooner or later, some kind of GUI structure will come up (you already have the text display if i see correctly).
Fortunatly, there are a lot of possibilities what can be done.

1. stay with the text display - you will find this is kind of limited.
2. stay with the text display and mirror it to the HDMI in text mode with more options only to see on the HDMI. (Better but not much)
3. don't bother about displays and just focus on external control over MIDI, OSC, etc. 
4. have a graphic display like the ILI9341 on the spi bus and mirror the content 1:1 to HDMI (don't know if that is possible with Circle, i think you would need some basic SDL graphic stuff to get this working) . This way, the display is an extra but everything can also be used on a regular PC monitor. 
5. The M8, dirty wave approach. https://dirtywave.com
This has a local display but also can send the display content as data via USB serial to a Webbrowser or to a small PC App that mirrors the display. This is partly the way i currenty also have done it for microdexed. (so far no webbrowser but a small (100 line) C Code, that can mirror the display 1:1 to a pc, mac, linux etc. over Serial USB. The performance penality is minimal, even on the teensy with it very limited ressources, compared to a Pie. Since the Pi also has Ethernet, that also could be a option to check. But even USB2 is more than enough, to handle this task. I think this is the best option. You can keep the Pie bare but also on the same time offer a standalone graphical UI, completly copied to a PC Monitor (or a Webbrowser - think of it, with a USB-WIFI Card this would even work with a phone or a tablet) and you have the identical codebase. 

If you choose something between 4-5, i think i could port over a lot of the GUI and sequencer stuff.

About the effects and instances, i am sure a Pi > 2 can handle this, probably even on 1 core.
The reverb in microdexed currently is basically this: https://github.com/hexeguitar/t40fx
It is a very good sounding reverb for its class on the teensy but very demanding on memory (on this platform). I believe you can port this to the Pi, if that not already is done - and memory on the Pi is "unlimited", in comparison.

**Reply by probonopd** [reply]
> Everything is deepy interwoven with the teensy audio library

https://codeberg.org/dcoredump/Synth_Dexed/ is being made a bit less tigtly coupled to the Teensy; see e.g., https://github.com/probonopd/MiniDexed/pull/41.

> sooner or later, some kind of GUI structure will come up (you already have the text display if i see correctly)

Well, the objective is to avoid a "computer" look and feel (after all, we can run Reaper on the RPi 4 and it excels at running Dexed and effects already). Currently we are using HDMI to basically output debug text information, and have added an DX7-style [inexpensive display](https://www.berrybase.de/sensoren-module/displays/alphanumerische-displays/alphanumerisches-lcd-16x2-gr-252-n/gelb) with a rotary encoder (so basically option 2 above).

> don't bother about displays and just focus on external control over MIDI, OSC, etc.

That's basically the idea for the next step. The idea is to implement TX802-style performance patches and sysex. This way everything could be done by MIDI.

> some basic SDL graphic stuff

Circle has [LVGL](https://lvgl.io/) support. [minisynth ](https://github.com/rsta2/minisynth) is an example using it on HDMI.

> Webbrowser

Yes, I think this might be a straightforward way to go, letting people use existing tablets and phones instead of having to buy a special screen and wire/solder much. Circle has HTTP server support and a BSD-like C++ socket API.

What I am wondering is whether we should try to get RTP-MIDI (MIDI-over-Ethernet) going. This way we could cleanly seperate the GUI from the engine had have everything controlled over MIDI, re-using the exsiting code. https://github.com/probonopd/MiniDexed/issues/43

**Reply by rsta2** [reply]
RTP-MIDI should be absolutely doable on the Raspberry Pi 2-4. I'm unsure, if the RPi 1 has enough performance to handle the network simultaneous to the sound rendering and USB. This would have to be tried.

**Reply by positionhigh** [reply]
I'd like to try out your build on the weekend and think i can follow your instructions how this is done, even i have no experience in your toolchain/environment. A few questions:
do i have to compile/build and then copy to the sd-card the binaries for every build or is there a way to directly access the Pi over USB to kind of boot it up from the development PC ? 
If that is not possible - can i compile and run the binary directly from an development environment that is running on the Pi itself?

**Reply by rsta2** [reply]
There is support for a serial bootloader, if you have a USB-to-serial adapter. Please see [this](https://github.com/rsta2/circle/blob/master/doc/bootloader.txt) for info on setting it up. There is also a bootloader, which uses TFTP or HTTP over Ethernet [here](https://github.com/rsta2/circle/tree/master/sample/38-bootloader).

I think there is an appropriate toolchain for AArch64 for 64-bit Raspberry Pi OS [here](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-a/downloads) (the AArch64 Linux hosted bare-metal target (aarch64-none-elf)). I haven't tried this by myself.

**Reply by positionhigh** [reply]
Hallo Rene,

ich rate mal du kommst auch aus Deutschland ? Ich denke ich habe einen 3.3v USB-to-serial Adapter und sehe mir das am Wochenende an, ich danke Dir.

Hello Rene,

i asume you are from Germany, too, if not please ignore my message above:-)
Think i have a 3.3.v usb-serial adapter and will look into it, thank you!

**Reply by positionhigh** [reply]
btw, maybe it is not so much directly interesting to your project but i do have a rough youtube playlist about my progress on microdexed on the teensy.
https://www.youtube.com/playlist?list=PLHTypoMU1QoGOXPli8bjR6MknPiQpubHl

**Reply by rsta2** [reply]
I did not know MicroDexed, before taking part on MiniDexed, but it seems to be a great project. We want to create something similar for the Raspberry Pi and use the power of this machine. Yes, I live in Germany.

**Reply by positionhigh** [reply]
Just a short feedback - following your example script i was able to compile for PI4 without problems. The only missing thing were the root certificates (maybe that is already in place in newer OS versions, i used Ubuntu 18.04 since i had a machine available with this version) "sudo apt-get install ca-certificates" solved the issue.

Would it be somehow possible to (cross)compile and run the binary on something like https://github.com/TriForceX/MiyooCFW/wiki/Miyoo-Mini ? That would be very interesting for me as a development platform since it already has a screen, lots of buttons, audio, headphones, sdcard, usb etc. and costs nearly nothing.

**Reply by probonopd** [reply]
I doubt Circle (which is the framework we are using) can run easily on anything but the various Raspberry Pis. We use that hardware because many people already have it.

**Reply by rsta2** [reply]
Circle runs on Raspberry Pis only. It's not planned to support other hardware.

**Reply by positionhigh** [reply]
tnx i thought that it might be very difficult but possible since the architecture seems at least partly similar. So to get my sequencer and interface things working with your Pi build i woud have to figure out if i can get a display like a ILI9341 working in this context. I looked at [LVGL](https://lvgl.io/) and it seems very fancy for simpe UIs but way too resource intense, also it does not seem to have any grid input/outputs so i think a simple, pixel based approch is still the way to go.
I understand that such things like a sequencer and on board GUI are not in your main focus. My goal is to get more to the features of https://dirtywave.com/ - Meaning multitimbral microdexed, also other synth engines, sampling, effects and sequencing in a standalone box with a touch screen or attached (USB) MIDI Controller.

**Reply by rsta2** [reply]
No, it's not that simple, and if you support other hardware, you also have to test this on any release. You need some focus and for Circle the focus is the Raspberry Pi.

Circle already supports SPI displays with the ST7789, if this helps. Programming with LVGL is special, but resources (memory) should not be a problem, because there is a lot in the Raspberry Pis. But LVGL does not work on SPI displays at the moment with Circle.

I'm not the project owner of MiniDexed, so I do not define the requirements, but supporting other synth engines and sampling seems to be a different project to me.

**Reply by probonopd** [reply]
Right now the project focus is on creating an instrument on Raspberry Pis that approximates the DX1/5/7 and TX802/816 plus mixer and effects. I'd like to keep the hardware barriers to entry really low (as in: low cost and easy access to the needed hardware) and have a "robust" instrument that boots fast and is very reliable.

**Reply by positionhigh** [reply]
totaly get it - focus on the basics first. I have so say, the fast boot is also one of my priorities - if you haven't gone the way with circles but instead booting up kind of a full linux, i would not be so much interested in the PI as a potential platform for a synth:_)
Since i can't help much on the low level stuff i will follow your progress and continue what i am doing on the teensy. Maybe in the future you will come up with some kind of "plugin" model where i can contribute my stuff as an option to your great project. All the best, cu soon!

**Comment by mrfloydst** [comment]
Hello probonopd,
thanks for reaching out! I've answered your mail but I'm afraid it went directly to the spam folder :-)
I'll post a video on your project, working on it right now.
And I'd like to join your effort by providing a manual for end users / musicians (and maybe a patch set ?)
I guess it's not possible to connect the LC display with the I2C interface mounted on it's back?

**Reply by probonopd** [reply]
Hello @mrfloydst, sorry for my late reply, went hunting for your mail when I saw your message here :-)

> I'll post a video on your project, working on it right now.

Very cool! Looking forward to it.

> And I'd like to join your effort by providing a manual for end users / musicians (and maybe a patch set ?)

This would be tremedous. There is some initial (short) documentation at https://github.com/probonopd/MiniDexed/wiki. Be sure to check out especially the "Menu" page, it explains the menu-based operation using the display and a rotary encoder.

> I guess it's not possible to connect the LC display with the I2C interface mounted on it's back?

Unfortunately the software doesn't support I2C yet, but you should still be able to use your display by soldering to the connections on the front, I guess (never tried but why shouldn't it work).

**Reply by probonopd** [reply]
The video has landed:
https://www.youtube.com/watch?v=Z3t94ceMHJo

Thank you very much @mrfloydst :+1:

**Reply by luisfcorreia** [reply]
it's basicly the reason why I'm here

**Reply by probonopd** [reply]
Also being discussed over at https://news.ycombinator.com/item?id=31062237

**Comment by positionhigh** [comment]
Hello minidexed friends,

just to keep you updated, the issues with the flash chip on the teensy front is now solved.

https://youtu.be/lwLht-6d_2o

**Comment by terracide303** [comment]
Would it be possible to add a HID driver so we can load devices that act like a HID keyboard but can be used to add buttons to shortcuts. I am talking about devices for retro gaming and arcade machines like this: https://www.highflow.nl/arcade/u-hid/ultimarc-u-hid-full-featured-version-pcb-with-usb-cable.html?sl=nl.    These can also be used to have hardware buttons just over USB. Think that is very handy? You could of course flash anything as a HID device but this was just an example.

**Reply by probonopd** [reply]
HID keyboards are already supported, but for some MIDI controllers (keyboards) one needs to force USB speed to full with `usbspeed=full` in `cmdline.txt` due to https://github.com/probonopd/MiniDexed/issues/48; and since most HID keyboards don't support that USB speed they won't work in conjunction. So unless this can be solved in [Circle](https://github.com/rsta2/circle/), I fear there is not much we can do in MiniDexed.

**Comment by 172pilot** [comment]
Just learned of this project today, and have my amazon cart full of parts to join in the fun!   I own an original DX7, so this will be a fun additional sound module for me.  I'm curious, and will probably just learn about this once I get my parts together, but if running more than one sound module, is there a way to pan different sounds/channels to different places in the L/R channels or is the whole thing mono like an original DX7?  Thanks for all the hard work.  Can't wait to start to play with this!!

**Reply by Banana71** [reply]
Hi 172pilot,

yes, you can set Pan for every Tonegenerator.
- "Pan Stereo panning Can be set graphically from left to right"
- 
Until your parts are delivered you can read something here:
https://github.com/probonopd/MiniDexed/wiki/Menu
The settings are explained under Menu tree

Peter

**Reply by 172pilot** [reply]
Thanks for the reply and link.. That looks great!  I can't wait!

**Comment by Fetente77** [comment]
Hi there, I built the minidexed and it's fantastic. I have, however, a pair of question: why it doesn't save modifies on shutdown? And, is there a way to create my own rom? Thanks all,

**Reply by positionhigh** [reply]
hi! if you have meant microdexed, please come to our discord chat, https://discord.gg/XCYk5P8GzF
much easier to ask questions and get quick replies.

If you want to save the current state, you have to save it as a performance. you have 99 performance slots.
and by your own rom you mean build it from source ? or with other rom samples ? both can be done, yes. however it is more convinient to load user samples from spi chip.

**Reply by Banana71** [reply]
Hi Fetente77,
I manage my banks with DEXED and I also edit the patches on the PC.
I create performances directly on the miniDexed and save them as "new performance" on microSD.
https://github.com/probonopd/MiniDexed/wiki/Menu    (scroll down to Performance)


![Dexed](https://user-images.githubusercontent.com/104296205/197580282-599a76fe-4e5d-403b-aace-f167b1de901c.jpg)

**Reply by Fetente77** [reply]
For positionhigh: I use minidexed for raspberry.

Many thanks for your reply. Now I’ve understood the meaning of performance.ini.
So if I install the dexed plugin, I can create my own syx rom and then import them on my sd?

**Comment by vicmertens** [comment]
I am starting to build synths an wondered if there is a manual available for the latest version of microdexed touch, with details for the sequencer?

**Comment by positionhigh** [comment]
It is not a full manual but you can find most information in the WIKI
https://codeberg.org/positionhigh/MicroDexed-touch/wiki/

**Comment by viragored** [comment]
Hi - Hello from New Zealand.
I've just put together a MiniDexed running on a Raspberry Pi 4, housed in a 3D printed case of your design. Excellent work everyone!
The most amazing thing to me is that one day I was Googling something to do with synths on R Pi, and saw for the first time mention of MiniDexed with the comment that it just works. Less than an hour later I was getting sound out of a R Pi 3 that I borrowed from something else!

And a question: 
I don't have much experience of building electronic things and R Pis so would welcome pointers to where I should be looking to get the information for the next step. 
I'd like to use a player program on my desktop R Pi 4 to play a midi file to the MiniDexed, and also plug in my Alesis keyboard/controller to play along with the playing midi tune or sequence. 
What are the magic ingredients I need to join together a midi file, a suitable player program, USB output from the desktop Pi 4, and USB input into MiniDexed on the other Pi 4? After spending ages in Google, I'm all at sea in terms I don't understand and need to take a break!
I hope someone can provide me with some links to help me.

PS: I've replaced the display in the photo with one that works properly!
![Working](https://github.com/probonopd/MiniDexed/assets/28804416/abbaba2a-5f11-4c9a-be11-59f85d8726d1)

**Reply by probonopd** [reply]
You can send MIDI data from the PC to your MiniDexed device using e.g., a Raspberry Pi Pico based USB to serial converter like explained [here](https://github.com/probonopd/MiniDexed/wiki/Hardware#usb-midi-devices).

Alternatively, on _some_ Raspberry Pi models, you might be able to use USB Gadget mode (no need for an additional device) as described [here](https://github.com/probonopd/MiniDexed/wiki/Hardware#usb-gadget-mode), but apparently that doesn't work with the Raspberry Pi 3 Model B.

**Reply by viragored** [reply]
Ah - a USB-serial converter! I've seen the term, didn't understand it, but now know what to look for. Thanks so much, 
Tomorrow I'll start Googling and hope to find something within my skills range.

**Reply by probonopd** [reply]
Personally I recommend the Raspberry Pi Pico (it's under 5 USD) and flash the software midi_adapter.uf2.zip software onto it from the link above to turn it into a USB-MIDI-to-serial converter that you can attach to the Raspberry Pi. It works for me with all Raspberry Pi models, and you can still keep your MIDI controller (keybord) attached to MiniDexed if you like.

**Reply by diyelectromusic** [reply]
You can run a Pi 4 in Gadget mode, using the USB-C port IIRC... but you need to be really careful how it is powered, as you might not get enough power from your host if powering the Pi that way...  Read the notes carefully on the wiki if you want to go this route so you don't damage your Pi or the thing you plug it into! (I see this as somewhat "advanced" for this reason).

I found the 3A+ the easiest "it just works" USB gadget device for MiniDexed, but the Zero (1 TG) or Zero 2 (if you want all 8 TGs) should be pretty straight forward too.

Aside: here's a USB to MIDI converter using a Seeed XIAO that can be built onto a PCB to plug directly into a Pi: https://diyelectromusic.wordpress.com/2023/04/16/xiao-pi-usb-midi-device-pcb/
(but you'd need to make your case a bit bigger!)

Kevin

**Reply by viragored** [reply]
Thanks so much for the help! I still feel like I'm swimming in a novice's sea of confusion, so perhaps you could tell me if I'm on the right lines here?
The picture shows what I think I'm aiming for. 
![USB Midi connections](https://github.com/probonopd/MiniDexed/assets/28804416/7498b483-6e37-4b13-8731-ee4e4d6ba5aa)
A constraint is that I don't want to change the way my desktop Pi 4 works - that needs to keep running with Pi OS and not try to be a USB Gadget. 
There's a clear description of how to make a Pico USB - Serial MIDI converter [here](https://diyelectromusic.wordpress.com/2021/01/23/midi-micropython-and-the-raspberry-pi-pico/), the breadboard picture is particularly helpful. 
It uses 3.3V, GND, and UART1, two resistors and a DIN socket - I think the physical side I can do.
Software may be another story. I found a link from [Adafruit](https://blog.adafruit.com/2022/04/21/a-low-cost-usb-to-serial-midi-converter-using-an-rp2040-and-circuitpython-raspberrypi-circuitpython-raspberry_pi-sandyjmacdonald/) to some CircuitPython [here](https://gist.github.com/sandyjmacdonald/2db20b2fecccd44b39d737fdc4d14d11) 
If I make the various numbers in the code match the physical pins, will that combination work?
And then I think I buy a cheap converter cable [such as this](https://www.trademe.co.nz/a/marketplace/music-instruments/pa-pro-audio-dj-equipment/computer-recording/listing/4550858911) to input to the MiniDexed Pi.
And if you have any links better than the ones I've found, I'd love to have them.
Thanks again!

**Reply by diyelectromusic** [reply]
I might be biased but there is a complete description of the Circuitpython code/approach here: https://diyelectromusic.wordpress.com/2022/04/21/circuitpython-usb-to-serial-midi-router/

I'm not clear if you want both the adaptor (USB) AND a serial MIDI interface to your MiniDexed?  I'm guessing don't have traditional MIDI on MiniDexed, but just want some kind of link?

Note: that first link is a MIDI OUT port.  MIDI IN is described [on the Wiki](https://github.com/probonopd/MiniDexed/wiki/Hardware#classic-midi-input).

So your options as I see them:
* PC direct USB MIDI connection - but will require something at the MiniDexed end that acts as a USB device/gadget.  There are several options here:
** If using a 3A+, 4 or Zero for your MiniDexed this could run in gadget mode itself and is by far the simplest (power issues mentioned above not withstanding).
** A device like the PIco solution midi_adaptor but you will need a way to connect this Pico to your MiniDexed.  The simplest connection is using the UART on the PIco to connect to the UART on the Pi.  This is basically "serial MIDI but without the serial MIDI circuit".  This is what my XIAO PCB is doing too and also the other Circuitpython based solutions.
** Something like two Picos back to back both acting as USB MIDI devices and connected by their own UARTs.  One Pico plugs into your PC over USB, one plugs into MiniDexed over USB.

* PC Serial MIDI connection using a USB to serial MIDI interface like the cheap one you show.  This requires a serial MIDI interface on the MiniDexed itself, which can be built in a DIY way (see the wiki) or could be a 3V3 ready bought MIDI module/adapter for the Pi.

* Use a keyboard MIDI controller that does serial MIDI to USB MIDI THRU so that is connected to both your PC and your MiniDexed.

You'll end up either with two USB devices plugged into MiniDexed (the PC via a DIY USB device adaptor; and the keyboard) or one serial MIDI (PC via a USB to serial MIDI adaptor) and one USB MIDI (the keyboard).  Either way you'll need something at the MiniDexed end to connect it to the PC.

Also note that some of those cheap USB MIDI adapters for PCs can be problematic if sending lots of data (http://www.arvydas.co.uk/2013/07/cheap-usb-midi-cable-some-self-assembly-may-be-required/).

Kevin

**Reply by viragored** [reply]
Thanks so much for the reply, Kevin. A penny just dropped! Making the MiniDexed a USB Gadget means my desktop Pi can plug straight in. The desktop peripherals are connected via a USB sharer which has additional power input, so I hope that will be enough to power the MiniDexed Pi 4 as a Gadget. I'll try that later this week. Sorry I had hold of the wrong end of the stick before!

**Reply by diyelectromusic** [reply]
There are quite a few options :)  Good luck!  Let us know how you get on!

**Reply by probonopd** [reply]
Other than on your picture, to send MIDI from the "PC" to the MiniDexed, you can use only one Raspberry Pi Pico and nothing else. This Raspberry Pi Pico plugs into your "PC" USB port on the one side, and into the RX/TX pins of your MiniDexed on the other side. Thus, it acts as the USB-MIDI-to-Serial converter. Hope this clears it up.

If you go the "USB gadget" route, then you won't need even the Raspberry Pi Pico, but it means that it will work only with certain models of MiniDexed devices, and you need to be careful to think about how to power your devices. As always, no guarantees.

**Reply by viragored** [reply]
Oh gosh, I have found it so hard to get my head round what's what with midi sources, synth sounds, and Pis generally. Your simple sentence certainly has cleared up some of it!! (If I've understood correctly what I think you're saying...)

Mission: Plug midi music file source into MiniDexed, plug keyboard into MiniDexed, play both 
Pico Hardware: Got it, I think.
Software: All confused... sorry. 

EDIT: I just noticed that the link on the wiki hardware page is to a .uf2 file. My bad for not spotting that earlier, sorry.
I've now got a Pico ready to flash and hook up to the MiniDexed. As soon as I get some time I'll be onto that!

Thanks once more for all the patient help.

**Reply by viragored** [reply]
A Pico is now loaded with the .uf2 file, hooked up to MiniDexed, all working!! Thanks so much for all the help and guidance.
Now I need to make a bigger box for the MiniDexed. (The project's design (my print is in my picture above) isn't too happy now with all those DuPont connectors, wires, floating boards and jacks and I fear something will get bent, broken or make an unwanted contact and do permanent damage).

**Reply by viragored** [reply]
With many thanks to everyone for all the help and support, here's a pic of my final (for now) setup.  
I've remixed a MiniDexed case on Thingiverse making it big enough to hold the Pi, LCD screen, rotary encoder/switch, DAC and voltage converter, they were a bit of a squeeze in the original case. Also inside now are connections for MIDI in and RCA sockets for audio output from the DAC. 
The little box behind the MiniDexed houses a Raspberry Pico with the .uf2 file that makes it a MIDI converter. It's connected to my desktop R Pi by USB, with output through a 3.5mm stereo jack plugged into the MiniDexed. The Alesis keyboard is plugged into the Pi's USB.
Now it's time to learn more about what's going on where and get some proper backing tracks running!
![MiniDexed complete 1](https://github.com/probonopd/MiniDexed/assets/28804416/38b355cb-39ef-4ebe-bce0-d6ad8d7c59bb)

**Reply by Banana71** [reply]
very nice work, a fully featured miniDexed.  I wish I could remove the protective film from the display.  😂

**Reply by viragored** [reply]
If I catch you taking my protective film off, expect a slap on the hand!
Leaving them on means when they do finally fall off, I've still got a pristine screen underneath. OK, that's not the only way I'm odd...

**Reply by Banana71** [reply]
everyone has their own quirks:

https://youtu.be/lrjwBvB7APY?si=rNX4lk70xg-cnKe4

**Comment by viragored** [comment]


**Reply by diyelectromusic** [reply]
> Fingers crossed it'll all work...

Ooo, err... not sure about that.  As I say, I've not really looking into it.  I hope that won't damage the Pi!  Might be worth studying the power section in the datasheet/part schematic to be sure you're happy with what is going on...

It's not really my area but I know USB-C can negotiate power in a much more complex way to previous standards...

Kevin

**Comment by Maximo68** [comment]
Hi, I wanted to know if there is a tutorial that explains how performances are created.
How do you create a code in .ini format with its parameters?
Is there any particular software?
Thank you.

**Reply by Banana71** [reply]
Floyd Steinberg once made a video about sound design with the miniDexed
https://youtu.be/KDx8iS59mpQ?si=Njd6GwPQxh5t3PzL

**Reply by Maximo68** [reply]
Hi Banana71, but how do you create your performances? 
Are the banks modified in .sys format or is the code formulated with some software?

**Reply by Banana71** [reply]
The only thing I use is my miniDexed and Dexed as an editor on the PC + MIDI controller. My patches are only on the PC. With Dexed I can transfer the patches to the miniDexed via MIDI and also edit them directly. I make the performance settings on the miniDexed. When I get the chance, I'll write a short manual or make a video of how I create the patches and performances.

**Reply by Maximo68** [reply]
Grazie per la risposta. Buon lavoro

**Comment by mbwum73** [comment]
My minidexed is working fine. However during boot the lcd screen shows a weird message. Does anyone know what it means and how to solve it?
![20250127_115357](https://github.com/user-attachments/assets/841417e2-0ee5-425e-b170-97c2bb741be5)

**Reply by probonopd** [reply]
This is not how it should look. It should say "MiniDexed" in the first row and should show the date (e.g., 20250102) followed by the short git hash (which seems to be working). I don't know why the beginning of the message is garbled on your screen. Are you using an i2s "backpack" on your display?

**Reply by mbwum73** [reply]
I built this one: https://diyelectromusic.com/2022/08/16/minidexed-raspberry-pi-io-board-part-3/
So no i2s "backpack"  on the display.
Besides from the weird boot screen it works just fine once it's fully booted.

**Reply by probonopd** [reply]
> So no i2s "backpack" on the display.

I haven't tested that configuration in quite a while. Hopefully we haven't broken it. Maybe some timing issue in the initialization phase or something like that.

**Reply by diyelectromusic** [reply]
I'm not sure why the screen would fail in a way that shows some text but then stop.  an you let us see the HDMI output of the boot?  I wonder if it is getting stuck somewhere...

**Comment by positionhigh** [comment]
@probonopd i am looking into porting my microdexed stuff to circle - to get starting i was trying to compile minidexed but mostly get into these errors, any advise ?

../include/arm_neon.h:7563:1: error: inlining failed in call to 'always_inline' 'vcvtq_n_s32_f32': target specific option mismatch
 7563 | vcvtq_n_s32_f32 (float32x4_t __a, const int __b)

tried arm toolchains  11,12,13.. but it does not seem to make any difference.

**Reply by probonopd** [reply]
Hi @positionhigh. Not sure what causes this error, I don't remember having seen it.

We are using these toolchains:
https://github.com/probonopd/MiniDexed/wiki/Development#building-locally

**Comment by anycolorsmith** [comment]
![photo_2025-05-09_10-47-47](https://github.com/user-attachments/assets/f798127c-29a8-41da-bfd4-4bee3cb71b18)
Just my lunch box dexed.

**Reply by probonopd** [reply]
Lovely.

**Reply by Banana71** [reply]
Bread and butter sounds, miniDexed-Edition 🥰

**Comment by sm7x7** [comment]
Hello there! Any possibility to add a lofi fx to the minidexed? Just simple one like an s&h lfo, and some bit reducer? TY!

**Comment by abux77** [comment]
First off this is an amazing Open Source project, I have been using it to learn more about creating PCBs and boxing it up. 

I breadboarded this up and got everything working using the GY-PCM5102 board, a 1602 LCD, an rotary encoder, pushbuttons and MIDI via USB. I then put it together on protoboard and got it working as well. 

I just made a PCB and I have some issues that I am trying to troubleshoot, I am am just wondering what other people have done to troubleshoot issues that you may have had?

This is the road I am taking, just wondering if there are some better ideas. I have loaded up Raspberry Pi OS, and I am trying to test everything in the terminal via SSH. 

I enabled I2C and used the command _sudo i2cdetect -y 1_ to look for an I2C address, I am getting nothing back I figure I messed something up there.

I tested USB MIDI with _aseqdump -p <port_name>_ and that works (didn't expect it not to). Does anyone know how to test MIDI via the UART though?

![IMG_6461](https://github.com/user-attachments/assets/5feee5ca-75ff-41dd-bc4f-ec99d86642e3)

![IMG_6464](https://github.com/user-attachments/assets/b0589db4-0ea7-4e26-8b48-c09108d36863)

**Reply by probonopd** [reply]
That is an amazingly clean looking build. Congrats! The only thing I would change is that I would do the LCD screws from the inside of the case for an even cleaner look (even though you are using nice screws).

As for i2c, the picture shows that your display is working, so either you already have i2c working, or your display doesn't use it so you don't need it.

If your board has no real MIDI circuitry, you need a device that connects to the serial port on the RPi and presents itself to the desktop computer as a MIDI device. This can be done, e.g., using a Raspberry Pi Pico and https://github.com/rsta2/pico/tree/main/midi_adapter.

**Reply by abux77** [reply]
Thanks for the feedback, I should have left a better description previously. My main board that connects the RPI, LCD, encoder and pushbuttons is working. The display is using the 4-bit parallel connection so no I2C.

<img width="1756" height="1234" alt="Main Board" src="https://github.com/user-attachments/assets/d347752e-803b-406a-ae8b-da0a2ce05189" />

The screws from the back would look better but my ultimate ambition is to house it in a metal Hammond enclosure. Tayda will made one off custom drilled and printed enclosures, but I made a 3D print first to check the fit.

The things that I need to check are on the Connector Board. I have experience with serial MIDI on the Raspberry Pi Pico via MicroPython, I guess I need to make a Python script to test it out, I was just wondering if there are any established Linux tools as I am not very familiar with Linux. The I2C address I was checking for was for the DAC, I am going to try PWM audio output to see if my MIDI is working and then I need to figure out what I messed up on my DAC.

<img width="1936" height="1362" alt="IO Board" src="https://github.com/user-attachments/assets/06b861e5-a3a2-4870-881f-94788e0af160" />

Here is how I am connecting the two boards with a ribbon cable.

![8993B34B-6FF5-4C7D-83BF-4AEC5A98D56A](https://github.com/user-attachments/assets/1df2b6f0-1b1f-482e-a562-e7a1dbacbe29)

Again, amazing project and documentation, you have set a high bar.

**Comment by TomSyl51** [comment]
Thank you to those who made this VERY NICE RPi DX7.  I designed a case for it and am ready to add it to my synth collection.  What a joy!
![IMG_6955](https://github.com/user-attachments/assets/129076f4-38eb-4e32-8c3f-0448a60a9024)
